lib/Common/CommonPal.h

```diff
@@ -611,19 +611,20 @@ HRESULT ULongMult(ULONG ulMultiplicand, ULONG ulMultiplier, ULONG* pulResult);
 template <class TryFunc, class FinallyFunc>
 void TryFinally(const TryFunc& tryFunc, const FinallyFunc& finallyFunc)
 {
-    bool hasException = true;
-    try
+    class FinallyObject
     {
-        tryFunc();
-        hasException = false;
-    }
-    catch(...)
-    {
-        finallyFunc(hasException);
-        throw;
-    }
-
-    finallyFunc(hasException);
+    public:
+        FinallyObject(const FinallyFunc& finallyFunc) : finallyFunc(finallyFunc), abnormalTermination(true) {}
+        ~FinallyObject() { finallyFunc(abnormalTermination); }
+
+        void SetHasNoAbnormalTermination() { abnormalTermination = false; }
+    private:
+        const FinallyFunc& finallyFunc;
+        bool abnormalTermination;
+    } finallyObject(finallyFunc);
```

Comment:
No, it won't if the dtor was called because of an exception.
Meghana, please add a big warning comment that the finallyFunc isn't allowed to trigger any exceptions.  This also means it can't probe the stack or allocate memory...

---
In reply to: [100933437](https://github.com/Microsoft/ChakraCore/pull/2519#discussion_r100933437) [](ancestors = 100933437)